longhaul/longhaul.md

```diff
@@ -0,0 +1,235 @@
+
+### Walk
+
+* Tests will use more iothub features (twin, methods, C2D, etc).
+
+* Fault injection tests will operate the same as simple D2C tests, but will launch as a seperate test scenario.
+
+* Simple randomization of these operations will be supported.  Overlapping of different operations won't need to happen here.
+
+* test launch automation will be increased so tests can be run in Docker containers in more of a bulk fashion
+
+
+### Run
+
+* More test scenarios will be supported, hopefully with the same test core running with different options.
+
+* Tests can be controlled via desired properties and C2D messages.  This includes operation cadence, failure properties, fault rates, and operation choices.
+
+* Maximum chaos is supported.  
+
+* DPS is used to provision devices to test
+
+* Tests are run with all manner of authenticatoin (x509, symmetric key, edge, etc)
+
+## Operations to test
+
+Operations will be phased in as follows.  These operation names are also used inside the various test structures below.
+
+| Operation        | Stage |
+|------------------|-------|
+| D2C              | crawl |
+| C2D              | walk  |
+| getTwin          | walk  |
+| updateTwin       | walk  |
+| receiveTwinPatch | walk  |
+| handleMethod     | walk  |
+| invokeMethod     | walk  |
+| DPS register     | run   |
+| blob upload      | run   |
+
+
+## Lonhaul configuration and progress
+
+* D2C telemetry will be used to send time-series data for deshboarding.
+
+* Reported propertis will be used to send static properties and current status.
+
+* Some of these data will be duplicated in telemetry and reported properties
+
+| mechanism           | data                                      | update frequency    |
+|---------------------|-------------------------------------------|---------------------|
+| D2C                 | progress                                  | once per second     |
+| reported properties | platform, sdk, testConfig, progres, stats | four times a minute |
+| desired properties  | testConfig (for updates - in run phase)   | service initiated   |
+| C2D                 | control messages                          | service initiated   |
+
+## Sample D2C message
+
+```
+"longhaulInfo": {
+    "progress": {
+        "status": "running",
+        "runtime": "00:01:07",
+        "heapUsed": "12.83",
+        "activeObjects": 5184,
+        "completeOpertions": 5001,
+        "outstandingOperations": 5,
+        "slowOperations": 15,
+    },
+}
+```
+
+## Sample reported properties
+
+```
+"longhaulInfo": {
+    "platform: {
+        "os": "linux",
+        "framworkVersion": "Node v12.16",
+        "heapSize": 15.4,
+    },
+    "sdk": {
+        "language": "node",
+        "version": "2.0.0",
+        "installSource": "npm",
+        "sourceRepo": "Azure/azure-iot-sdk-node",
+        "sourceBranch": "master",
+        "sourePr": "512",
+        "sourceCommit": "a489ab11e53b9c0cfe5c8da9f048c98ae91e9d98",
+    },
```

Comment:
Having the sdk information in the twin is nice.  Would be able to get all this information during the longhaul test?  The github information would be really good, but is it easy to get?

Reply:
the difficulty of getting sdk and github info remains to be seen, but it should be possible in some scenarios.

---
In reply to: [441145912](https://github.com/Azure/azure-iot-sdk-python/pull/595#discussion_r441145912) [](ancestors = 441145912)